File: examples/auto-responder-scripts/distance.py
#!/usr/bin/env python3
# mm_meta:
#   name: Distance Calculator
#   emoji: 📏
#   language: Python
"""
Distance Calculator Script - Uses location environment variables.

This script demonstrates how to use the FROM_LAT/FROM_LON and MM_LAT/MM_LON
environment variables to calculate the distance between the sender and the
MeshMonitor node.

Requirements:
- Python 3.6+
- No external dependencies

Setup:
1. Copy this script to your scripts directory
2. Configure trigger in MeshMonitor UI:
   - Trigger Pattern: distance, dist
   - Response Type: Script
   - Response: /data/scripts/distance.py

Usage:
- distance - Calculates distance from sender to MeshMonitor node
- dist - Alias for distance

Environment Variables Used:
- FROM_LAT, FROM_LON: Sender's location
- MM_LAT, MM_LON: MeshMonitor node location
- FROM_NODE: Sender's node number (for display)
"""

import os
import json
import math
from typing import Optional, Tuple


def haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate the great-circle distance between two points on Earth.

    Args:
        lat1, lon1: First point (latitude, longitude in degrees)
        lat2, lon2: Second point (latitude, longitude in degrees)

    Returns:
        Distance in kilometers
    """
    R = 6371  # Earth's radius in km

    # Convert to radians
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)

    # Haversine formula
    a = math.sin(dlat / 2) ** 2 + \
        math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(dlon / 2) ** 2
    c = 2 * math.asin(math.sqrt(a))

    return R * c


def km_to_miles(km: float) -> float:
    """Convert kilometers to miles."""
    return km * 0.621371


def get_bearing(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate the bearing from point 1 to point 2.

    Args:
        lat1, lon1: Starting point (latitude, longitude in degrees)
        lat2, lon2: Ending point (latitude, longitude in degrees)

    Returns:
        Bearing in degrees (0-360)
    """
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    dlon = math.radians(lon2 - lon1)

    x = math.sin(dlon) * math.cos(lat2_rad)
    y = math.cos(lat1_rad) * math.sin(lat2_rad) - \
        math.sin(lat1_rad) * math.cos(lat2_rad) * math.cos(dlon)

    bearing = math.degrees(math.atan2(x, y))
    return (bearing + 360) % 360


def bearing_to_direction(bearing: float) -> str:
    """Convert bearing to compass direction."""
    directions = ["N", "NE", "E", "SE", "S", "SW", "W", "NW"]
    index = round(bearing / 45) % 8
    return directions[index]


def get_location(lat_var: str, lon_var: str) -> Optional[Tuple[float, float]]:
    """Get location from environment variables."""
    lat = os.environ.get(lat_var)
    lon = os.environ.get(lon_var)

    if lat and lon:
        try:
            return (float(lat), float(lon))
        except ValueError:
            return None
    return None


def main():
    """Main function to calculate and report distance."""
    try:
        # Get locations from environment
        from_loc = get_location("FROM_LAT", "FROM_LON")
        mm_loc = get_location("MM_LAT", "MM_LON")
        from_node = os.environ.get("FROM_NODE", "?")
        short_name = "!{}".format("{0:04x}".format(int(from_node))[4:])

        # Check if both locations are available
        if not from_loc:
            output = {
                "response": f"Hi {short_name}, your location is not available. "
                           "Send a position update first."
            }
            print(json.dumps(output))
            return

        if not mm_loc:
            output = {
                "response": f"Hi {short_name}, MeshMonitor location not set. "
                           "The node needs a GPS fix or manual position."
            }
            print(json.dumps(output))
            return

        # Calculate distance
        dist_km = haversine(from_loc[0], from_loc[1], mm_loc[0], mm_loc[1])
        dist_mi = km_to_miles(dist_km)

        # Calculate bearing (from sender to MM)
        bearing = get_bearing(from_loc[0], from_loc[1], mm_loc[0], mm_loc[1])
        direction = bearing_to_direction(bearing)

        # Format response based on distance
        if dist_km < 1:
            # Less than 1 km - show in meters
            dist_m = dist_km * 1000
            response = f"Hi {short_name}, distance: {dist_m:.0f}m ({direction})"
        elif dist_km < 10:
            # Less than 10 km - show one decimal
            response = f"Hi {short_name}, distance: {dist_km:.1f}km / {dist_mi:.1f}mi ({direction})"
        else:
            # Larger distances - show whole numbers
            response = f"Hi {short_name}, distance: {dist_km:.0f}km / {dist_mi:.0f}mi ({direction})"

        output = {"response": response}
        print(json.dumps(output))

    except Exception as e:
        error_msg = f"Error calculating distance: {str(e)}"
        if len(error_msg) > 195:
            error_msg = error_msg[:192] + "..."
        output = {"response": error_msg}
        print(json.dumps(output))
        print(f"Script error: {e}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: src/utils/distance.ts
/**
 * Distance calculation utilities for Meshtastic nodes
 */

/**
 * Calculate the great circle distance between two points on Earth using the Haversine formula
 * @param lat1 Latitude of point 1 in degrees
 * @param lon1 Longitude of point 1 in degrees
 * @param lat2 Latitude of point 2 in degrees
 * @param lon2 Longitude of point 2 in degrees
 * @returns Distance in kilometers
 */
export function calculateDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371; // Earth's radius in kilometers
  const dLat = toRadians(lat2 - lat1);
  const dLon = toRadians(lon2 - lon1);

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRadians(lat1)) * Math.cos(toRadians(lat2)) *
    Math.sin(dLon / 2) * Math.sin(dLon / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  const distance = R * c;

  return distance;
}

/**
 * Convert degrees to radians
 */
function toRadians(degrees: number): number {
  return degrees * (Math.PI / 180);
}

/**
 * Convert kilometers to miles
 */
export function kmToMiles(km: number): number {
  return km * 0.621371;
}

/**
 * Format distance for display based on user preference
 * @param km Distance in kilometers
 * @param unit 'km' or 'mi'
 * @param decimals Number of decimal places (default: 1)
 * @returns Formatted string with unit
 */
export function formatDistance(km: number, unit: 'km' | 'mi' = 'km', decimals: number = 1): string {
  if (unit === 'mi') {
    const miles = kmToMiles(km);
    return `${miles.toFixed(decimals)} mi`;
  }
  return `${km.toFixed(decimals)} km`;
}

/**
 * Node position interface for distance calculations
 */
interface NodeWithPosition {
  user?: { id?: string };
  position?: {
    latitude?: number;
    longitude?: number;
  };
}

/**
 * Calculate and format distance from home node to target node
 * @param homeNode The home/local node with position data
 * @param targetNode The target node to calculate distance to
 * @param unit Distance unit preference ('km' or 'mi')
 * @returns Formatted distance string or null if positions unavailable
 */
export function getDistanceToNode(
  homeNode: NodeWithPosition | undefined,
  targetNode: NodeWithPosition,
  unit: 'km' | 'mi'
): string | null {
  // Check if home node has valid position (use != null to allow 0 coordinates)
  if (homeNode?.position?.latitude == null || homeNode?.position?.longitude == null) return null;
  // Check if target node has valid position
  if (targetNode.position?.latitude == null || targetNode.position?.longitude == null) return null;
  // Don't show distance to self
  if (homeNode.user?.id && homeNode.user.id === targetNode.user?.id) return null;

  const km = calculateDistance(
    homeNode.position.latitude,
    homeNode.position.longitude,
    targetNode.position.latitude,
    targetNode.position.longitude
  );
  return formatDistance(km, unit);
}

/**
 * Format a human-readable accuracy estimate for a given Meshtastic position precision value.
 * Meshtastic encodes positions as int32 (1 unit = 1e-7 degrees). With N precision bits,
 * the lower (32-N) bits are zeroed, giving a grid cell of 2^(32-N) * 1e-7 degrees.
 * The accuracy shown is half the grid cell (max deviation from true position),
 * matching the values in the Meshtastic documentation.
 * @param bits Precision bits (0-32). 0 = disabled, 32 = full precision (~1 cm)
 * @param unit 'km' for metric (m/km) or 'mi' for imperial (ft/mi)
 * @returns Human-readable accuracy string like "~100 m", "~1.5 km", "~300 ft", "~2 mi"
 */
export function formatPrecisionAccuracy(bits: number, unit: 'km' | 'mi'): string {
  if (bits <= 0) return 'Disabled';

  const METERS_PER_DEGREE = 111111;
  const FEET_PER_METER = 3.28084;
  const FEET_PER_MILE = 5280;

  // Half the grid cell size = max deviation from true position
  const accuracyMeters = Math.pow(2, 32 - bits) * 1e-7 * METERS_PER_DEGREE / 2;

  if (unit === 'mi') {
    const feet = accuracyMeters * FEET_PER_METER;
    if (feet < 1) {
      return '< 1 ft';
    }
    if (feet < FEET_PER_MILE / 10) { // 528ft = 0.1 mile
      return `~${Math.round(feet)} ft`;
    }
    const miles = feet / FEET_PER_MILE;
    if (miles < 10) {
      return `~${miles.toFixed(1)} mi`;
    }
    return `~${Math.round(miles)} mi`;
  }

  // Metric
  if (accuracyMeters < 1) {
    return '< 1 m';
  }
  if (accuracyMeters < 1000) {
    return `~${Math.round(accuracyMeters)} m`;
  }
  const km = accuracyMeters / 1000;
  if (km < 10) {
    return `~${km.toFixed(1)} km`;
  }
  return `~${Math.round(km)} km`;
}
